BearlyHumans/squirrelville
# Completely changing the entire movement system

Trello cards:
- https://trello.com/c/uDC8D0Ir/118-add-special-jumps-over-gaps-and-around-corners
- https://trello.com/c/cymnGmMK/98-refine-squirrel-movement
- https://trello.com/c/cVKjAC5H/148-add-stamina-system-w-good-controls

Adds:
- Edge Detect
- Up-force elimination when rotating (improves climbing significantly)
- Corner mounting
- Stamina
- Dash for auto climb (early version)
- Bug fixes (e.g. clipping underneath ground)

Testing Steps:
- Ensure that the ground in your scene is in the Ground Layer.
- Tag different objects with "CanClimb", "HardClimb", "NoClimb" or no tag for EZClimb (Tags can be changed in the movement script inspector.

Careful Mode:
- Press space when not near a climbable object = the player should jump.
- Press space near a climable object = the player should start climbing it.
- Run into a sharp angle (close to 90deg, e.g. the side of a cube) = the player should stop at the edge.
- While moving towards the edge, press space = the player should jump around the corner.
- Run towards the same edge but at an angle = the player should slide along the edge and stay on the object.

Dashing Mode:
While holding shift:
- Run towards a climable object = the player should start climbing it.
- Run towards a similar sharp angle to before = the player should automatically jump around the corner.

Stamina:
- Hold shift (dash) and move = the green stamina circle on the screen should lose segments.
- Climb on objects which have been tagged with different climb difficulties:
    * Stamina goes down when on CanClimb.
    * Stamina goes down faster when on HardClimb.
    * Stamina doesn't go down when on untagged (EZClimb).
    * Stamina instantly goes to zero on NoClimb.
- Try to climb past the max Angle (default 130 deg, upside-down is 180) = player falls off the surface.

Known bugs:
- Edge detect is unreliable on non-90deg corners
- Corner Jumping occasionally fails to trigger, and a normal jump occurs instead.
